### PR TITLE
Eth to elrond flow integration part1

### PR DESCRIPTION
**Pull request by iulianpascalau**
- integration new flows & new clients - part 1

### CODE REVIEW

File: clients/roleProviders/elrondRoleProvider_test.go [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+func TestNewElrondRoleProvider(t *testing.T) {
+	t.Parallel()
+
+	t.Run("nil data getter should error", func(t *testing.T) {
+		args := createElrondMockArgs()
+		args.DataGetter = nil
+
+		erp, err := NewElrondRoleProvider(args)
+		assert.True(t, check.IfNil(erp))
```
Comment (sstanculeanu) [inline clients/roleProviders/elrondRoleProvider_test.go:30]: why not `assert.Nil(t, erp)` ?

File: clients/roleProviders/elrondRoleProvider_test.go [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+		assert.Equal(t, ErrNilDataGetter, err)
+	})
+	t.Run("nil logger should error", func(t *testing.T) {
+		args := createElrondMockArgs()
+		args.Log = nil
+
+		erp, err := NewElrondRoleProvider(args)
+		assert.True(t, check.IfNil(erp))
```
Comment (sstanculeanu) [inline clients/roleProviders/elrondRoleProvider_test.go:38]: same `assert.Nil(t, erp)`

File: clients/roleProviders/elrondRoleProvider_test.go [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+		assert.Equal(t, ErrNilLogger, err)
+	})
+	t.Run("should work", func(t *testing.T) {
+		args := createElrondMockArgs()
+
+		erp, err := NewElrondRoleProvider(args)
+		assert.False(t, check.IfNil(erp))
```
Comment (sstanculeanu) [inline clients/roleProviders/elrondRoleProvider_test.go:45]: `assert.NotNil(t, erp)`

File: clients/roleProviders/ethereumRoleProvider.go [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+// NewEthereumRoleProvider creates a new ethereum role provider instance able to fetch the
+// whitelisted addresses and able to check ethereum signatures
+func NewEthereumRoleProvider(args ArgsEthereumRoleProvider) (*ethereumRoleProvider, error) {
+	err := checkEthereumRoleProviderSpecificArgs(args)
+	if err != nil {
+		return nil, err
+	}
+
+	erp := &ethereumRoleProvider{
+		whitelistedAddresses:    make(map[common.Address]struct{}),
```
Comment (sstanculeanu) [inline clients/roleProviders/ethereumRoleProvider.go:38]: just a question for myself: should we keep the order of the parameters here same as in the struct?

**Review by sstanculeanu — COMMENTED** [review]
small considerations

File: clients/elrond/dataGetter_test.go [hunk 5]
```diff
@@ -823,3 +889,164 @@ func TestDataGetter_GetActionIDForSetStatusOnPendingTransfer(t *testing.T) {
...
+	proxyCalled := false
+	args.Proxy = &interactors.ElrondProxyStub{
+		ExecuteVMQueryCalled: func(ctx context.Context, vmRequest *data.VmValueRequest) (*data.VmValuesResponseData, error) {
+			proxyCalled = true
+			assert.Equal(t, args.RelayerAddress.AddressAsBech32String(), vmRequest.CallerAddr)
+			assert.Equal(t, args.MultisigContractAddress.AddressAsBech32String(), vmRequest.Address)
+			assert.Equal(t, "", vmRequest.CallValue)
+			assert.Equal(t, quorumReachedFuncName, vmRequest.FuncName)
+
+			expectedArgs := []string{hex.EncodeToString(big.NewInt(112233).Bytes())}
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply clients/elrond/dataGetter_test.go:906]: right, done

File: clients/elrond/dataGetter_test.go [hunk 6]
```diff
@@ -823,3 +889,164 @@ func TestDataGetter_GetActionIDForSetStatusOnPendingTransfer(t *testing.T) {
...
+		ExecuteVMQueryCalled: func(ctx context.Context, vmRequest *data.VmValueRequest) (*data.VmValuesResponseData, error) {
+			proxyCalled = true
+			assert.Equal(t, args.RelayerAddress.AddressAsBech32String(), vmRequest.CallerAddr)
+			assert.Equal(t, args.MultisigContractAddress.AddressAsBech32String(), vmRequest.Address)
+			assert.Equal(t, "", vmRequest.CallValue)
+			assert.Equal(t, signedFuncName, vmRequest.FuncName)
+
+			expectedArgs := []string{
+				hex.EncodeToString(args.RelayerAddress.AddressBytes()),
+				hex.EncodeToString(big.NewInt(112233).Bytes()),
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply clients/elrond/dataGetter_test.go:1003]: done

File: clients/roleProviders/elrondRoleProvider_test.go (hunk 1 above)
Comment (iulianpascalau) [reply clients/roleProviders/elrondRoleProvider_test.go:30]: Through this, I've also checked the IsInterfaceNil works properly :)

File: clients/roleProviders/ethereumRoleProvider.go (hunk 4 above)
Comment (iulianpascalau) [reply clients/roleProviders/ethereumRoleProvider.go:38]: order is not important, for now, in this component.
Reason: this role provider is used to determine if a certain signature received from the p2p subsystem is valid or not, ethereum-wise. That means it does several checks like: public key recovery, signature verification and public key authorization - which is a search in a map fed by data from the ethereum contract

File: clients/elrond/dataGetter.go [hunk 7]
```diff
@@ -305,6 +310,44 @@ func (dg *elrondClientDataGetter) GetActionIDForSetStatusOnPendingTransfer(ctx c
...
+}
+
+// GetLastExecutedEthTxID returns the last executed Ethereum deposit ID
+func (dg *elrondClientDataGetter) GetLastExecutedEthTxID(ctx context.Context) (uint64, error) {
+	builder := dg.createDefaultVmQueryBuilder().Function(getLastExecutedEthTxId)
+
+	return dg.executeQueryUint64FromBuilder(ctx, builder)
+}
+
+// WasSigned returns true if the action ID was signed by the current relayer
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply clients/elrond/dataGetter.go:335]: done

File: clients/roleProviders/ethereumRoleProvider.go (hunk 4 above)
Comment (sstanculeanu) [reply clients/roleProviders/ethereumRoleProvider.go:38]: understood. thanks

File: clients/roleProviders/elrondRoleProvider_test.go (hunk 1 above)
Comment (sstanculeanu) [reply clients/roleProviders/elrondRoleProvider_test.go:30]: correct. nice

File: clients/roleProviders/elrondRoleProvider.go [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+
+func (erp *elrondRoleProvider) processResults(results [][]byte) {
+	currentList := make([]string, 0, len(results))
+
+	erp.mut.Lock()
+	erp.whitelistedAddresses = make(map[string]struct{})
+
+	for _, result := range results {
+		address := data.NewAddressFromBytes(result)
+		currentList = append(currentList, address.AddressAsBech32String())
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply clients/roleProviders/elrondRoleProvider.go:74]: good suggestion :+1:

[the remark this answers is not in the document]
Comment (iulianpascalau) [reply clients/roleProviders/elrondRoleProvider.go:74]: done

File: clients/roleProviders/elrondRoleProvider.go [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+	if err != nil {
+		return err
+	}
+
+	erp.processResults(results)
+
+	return nil
+}
+
+func (erp *elrondRoleProvider) processResults(results [][]byte) {
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply clients/roleProviders/elrondRoleProvider.go:66]: done

File: clients/roleProviders/elrondRoleProvider_test.go [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+		[]byte("address 1"),
+		[]byte("address 2"),
+	}
+
+	t.Run("nil whitelisted", testElrondExecuteShouldWork(nil))
+	t.Run("empty whitelisted", testElrondExecuteShouldWork(make([][]byte, 0)))
+	t.Run("with whitelisted", testElrondExecuteShouldWork(whitelistedAddresses))
+}
+
+func testElrondExecuteShouldWork(whitelistedAddresses [][]byte) func(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply clients/roleProviders/elrondRoleProvider_test.go:79]: done

File: clients/roleProviders/elrondRoleProvider_test.go [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+	return ArgsElrondRoleProvider{
+		Log:        logger.GetOrCreate("test"),
+		DataGetter: &bridgeV2.DataGetterStub{},
+	}
+}
+
+func TestNewElrondRoleProvider(t *testing.T) {
+	t.Parallel()
+
+	t.Run("nil data getter should error", func(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply clients/roleProviders/elrondRoleProvider_test.go:25]: done

File: clients/roleProviders/elrondRoleProvider_test.go [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+		args := createElrondMockArgs()
+		args.DataGetter = nil
+
+		erp, err := NewElrondRoleProvider(args)
+		assert.True(t, check.IfNil(erp))
+		assert.Equal(t, ErrNilDataGetter, err)
+	})
+	t.Run("nil logger should error", func(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply clients/roleProviders/elrondRoleProvider_test.go:33]: done

File: clients/roleProviders/elrondRoleProvider_test.go [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+		args := createElrondMockArgs()
+		args.Log = nil
+
+		erp, err := NewElrondRoleProvider(args)
+		assert.True(t, check.IfNil(erp))
+		assert.Equal(t, ErrNilLogger, err)
+	})
+	t.Run("should work", func(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply clients/roleProviders/elrondRoleProvider_test.go:41]: done

File: clients/roleProviders/ethereumRoleProvider_test.go [hunk 14]
```diff
@@ -0,0 +1,151 @@
...
+	return ArgsEthereumRoleProvider{
+		Log:                     logger.GetOrCreate("test"),
+		EthereumChainInteractor: &interactors.EthereumChainInteractorStub{},
+	}
+}
+
+func TestNewEthereumRoleProvider(t *testing.T) {
+	t.Parallel()
+
+	t.Run("nil ethereum chain interactor should error", func(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply clients/roleProviders/ethereumRoleProvider_test.go:27]: done

File: clients/roleProviders/ethereumRoleProvider_test.go [hunk 15]
```diff
@@ -0,0 +1,151 @@
...
+		args := createEthereumMockArgs()
+		args.EthereumChainInteractor = nil
+
+		erp, err := NewEthereumRoleProvider(args)
+		assert.True(t, check.IfNil(erp))
+		assert.Equal(t, ErrNilEthereumChainInteractor, err)
+	})
+	t.Run("nil logger should error", func(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply clients/roleProviders/ethereumRoleProvider_test.go:35]: done

File: clients/roleProviders/ethereumRoleProvider_test.go [hunk 16]
```diff
@@ -0,0 +1,151 @@
...
+		args := createEthereumMockArgs()
+		args.Log = nil
+
+		erp, err := NewEthereumRoleProvider(args)
+		assert.True(t, check.IfNil(erp))
+		assert.Equal(t, ErrNilLogger, err)
+	})
+	t.Run("should work", func(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply clients/roleProviders/ethereumRoleProvider_test.go:43]: done

File: clients/roleProviders/ethereumRoleProvider_test.go [hunk 17]
```diff
@@ -0,0 +1,151 @@
...
+		common.HexToAddress("0xb6e20FF4Ae7d29be233D874633F2F0Dcb326E5c0"),
+	}
+
+	t.Run("nil whitelisted", testEthereumExecuteShouldWork(nil))
+	t.Run("empty whitelisted", testEthereumExecuteShouldWork(make([]common.Address, 0)))
+	t.Run("with whitelisted", testEthereumExecuteShouldWork(whitelistedAddresses))
+}
+
+func testEthereumExecuteShouldWork(whitelistedAddresses []common.Address) func(t *testing.T) {
+	return func(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply clients/roleProviders/ethereumRoleProvider_test.go:82]: done

File: clients/roleProviders/ethereumRoleProvider_test.go [hunk 18]
```diff
@@ -0,0 +1,151 @@
...
+	whitelistedAddresses = []common.Address{
+		common.HexToAddress("0x132A150926691F08a693721503a38affeD18d524"),
+	}
+	hexMsg = "c124f221e1992619dfe3254e46a97bd7d787ed4e699f48aca715d54e7f52ff5d"
+	hexSig = "b0ddb854c7c6a5c78cdbf9e7e6c204711c162220298dc1bfab58be77b8627c155ae4dac5d06197283407993b359752f8906487fc0e3a031173fd07c010e5cddc00"
+	t.Run("address not whitelisted", testEthereumVerifySigShouldWork(whitelistedAddresses, hexSig, hexMsg, ErrAddressIsNotWhitelisted))
+}
+
+func testEthereumVerifySigShouldWork(whitelistedAddresses []common.Address, hexSig string, hexMsg string, expectedErr error) func(t *testing.T) {
+	return func(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply clients/roleProviders/ethereumRoleProvider_test.go:126]: done

File: factory/ethElrondBridgeComponents.go [hunk 19]
```diff
@@ -0,0 +1,334 @@
...
+	logger "github.com/ElrondNetwork/elrond-go-logger"
+	erdgoCore "github.com/ElrondNetwork/elrond-sdk-erdgo/core"
+	"github.com/ElrondNetwork/elrond-sdk-erdgo/data"
+	"github.com/ElrondNetwork/elrond-sdk-erdgo/interactors"
+	"github.com/ethereum/go-ethereum/common"
+	ethCrypto "github.com/ethereum/go-ethereum/crypto"
+)
+
+const (
+	// minimumDurationForStep          = time.Second
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply factory/ethElrondBridgeComponents.go:34]: should have been used in the next PR, removed

**Review by sstanculeanu — APPROVED** [review]

**Review by dragos-rebegea — APPROVED** [review]

**Pull request merged**